apps/qubit/modules/informationobject/actions/renameAction.class.php

```diff
@@ -57,20 +68,41 @@ private function updateFields()
       $resource->title = $postData['title'];
     }
 
-    // Update slug, if requested
+    // Attempt to update slug if slug sent
     if (isset($postData['slug']))
     {
       $slug = QubitSlug::getByObjectId($resource->id);
-      $slug->slug = $postData['slug'];
-      $slug->save();
+
+      // Attempt to change slug if submitted slug's different than current slug
+      if ($postData['slug'] != $slug->slug)
+      {
+        $slug->slug = InformationObjectSlugPreviewAction::determineAvailableSlug($postData['slug']);
+        $slug->save();
+      }
     }
 
     // Update digital object filename, if requested
     if (isset($postData['filename']) && count($resource->digitalObjects))
     {
+      // Parse filename so special characters can be removed
+      $fileParts = pathinfo(trim($postData['filename']));
+      $filename = QubitSlug::slugify($fileParts['filename']) .'.'. QubitSlug::slugify($fileParts['extension']);
+
       $digitalObject = $resource->digitalObjects[0];
-      $digitalObject->name = $postData['filename'];
+
+      // Rename master file
+      $basePath = sfConfig::get('sf_web_dir') . $digitalObject->path;
+      $oldFilePath = $basePath . DIRECTORY_SEPARATOR . $digitalObject->name;
+      $newFilePath = $basePath . DIRECTORY_SEPARATOR . $filename;
+      rename($oldFilePath, $newFilePath);
+      chmod($newFilePath, 0644);
+
+      // Change name in database
+      $digitalObject->name = $filename;
       $digitalObject->save();
+
+      // Regeneraate derivatives
+      digitalObjectRegenDerivativesTask::regenerateDerivatives($digitalObject);
```

Comment:
I think we should check if they exist and delete the old derivates before. I guess it's not done in the regenerateDerivatives task when the filename is changed.